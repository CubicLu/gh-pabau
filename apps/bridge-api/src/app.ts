import express from 'express'
import cors from 'cors'
import { server } from './server'

interface Options {
  tracing: boolean
  debugApollo: boolean
  logging: boolean
}

/**
 * Creates a testable Express app
 */
export function createApp(options?: Options) {
  const app = express()
  app.set('trust proxy', true)
  app
    .get('/', function ({ res }) {
      res.send('<h1>Private</h1>')
    })
    .use(cors())
    .use(
      express.urlencoded({
        extended: true,
      })
    )

  server.applyMiddleware({ app })
  return app
}
