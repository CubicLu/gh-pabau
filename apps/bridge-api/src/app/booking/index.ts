export * from './nexus-type'
export * from './queries'
export * from './mutations'
