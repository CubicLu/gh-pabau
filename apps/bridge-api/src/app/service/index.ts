export * from './queries'
export * from './nexus-type/index'
export * from './nexus-type/Public_ServiceUserTier'
