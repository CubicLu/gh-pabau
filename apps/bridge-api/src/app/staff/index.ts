export * from './nexus-type'
export * from './queries'
