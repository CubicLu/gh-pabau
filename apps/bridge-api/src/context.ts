import jwt from 'jsonwebtoken'
import { prisma } from './prisma'
import { ExpressContext } from 'apollo-server-express'
import { version } from '../../../package.json'
import { ContextFunction } from 'apollo-server-core'
import { PrismaClient } from '@prisma/client'
import { JwtUser } from '@pabau/yup'

export interface Context {
  /**
   * The Prisma Client
   */
  prisma: PrismaClient
  /**
   * For Authentication module to resolve pod user.id, we have this emergency hatch that returns a PrismaClient instance
   * for the specified pod.
   *
   * @param remote_url - The `remote_url` from the database
   */

  prismaArray(remote_url: string): PrismaClient

  /**
   * The currently logged in user
   */
  authenticated?: JwtUser

  /**
   * The package.json version for this app
   */
  version: string

  /**
   * requested JWT to forward via Third Party services (eg. CDN)
   */
  authJwt: string

  /**
   * Required for PrismaSelect plugin to operate
   */
  select: any
}

export const createContext: ContextFunction<ExpressContext, Context> = (
  req
) => {
  const ret = {
    version,
    select: {},
  } as Context
  const authorizationRaw = req.req.header('authorization')
  if (authorizationRaw) {
    try {
      ret.authenticated = jwt.verify(
        authorizationRaw.replace(/^Bearer /, ''),
        process.env.JWT_SECRET,
        { algorithms: ['HS512'] }
      ) as JwtUser
    } catch {
      console.log('invalid jwt found')
    }
  }
  ret.authJwt = authorizationRaw
  if (
    !ret.authenticated &&
    req.req.header('remoteurl') &&
    req.req.header('remoteurl') !== 'null'
  ) {
    ret.prisma = prisma(req.req.header('remoteurl'))
  } else {
    ret.prisma = prisma(ret.authenticated?.remote_url || undefined)
  }

  ret.prismaArray = (remote_url) => prisma(remote_url)

  return ret
}
