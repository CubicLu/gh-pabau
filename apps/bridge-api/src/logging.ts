// thanks to https://github.com/apollographql/apollo-server/issues/4055
export const BASIC_LOGGING = {
  requestDidStart() {
    return {
      didEncounterErrors(requestContext) {
        console.error(
          'an error happened in response to query ' +
            requestContext.request.query
        )
        console.log(requestContext.errors)
      },
      willSendResponse(requestContext) {
        console.log(
          'response sent',
          JSON.stringify(requestContext?.response?.data)
        )
        console.log(requestContext.response.http)
      },
    }
  },
}
