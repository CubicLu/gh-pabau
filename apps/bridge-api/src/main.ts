import {
  assertEnvVarsExist,
  stringToBoolean,
  truthyToMaskedString,
} from './utils'
import { config } from 'dotenv-flow'
import { version } from '../../../package.json'
import { createApp } from './app'

console.log(`Starting bridge-api v${version}`)

const LOGGING = !!stringToBoolean(process.env['LOGGING'])
const TRACING = !!stringToBoolean(process.env['TRACING'])
const DEBUG_APOLLO = !!stringToBoolean(process.env['DEBUG_APOLLO'])
const PORT = process.env['PORT'] || 4000

createApp({
  tracing: TRACING,
  logging: LOGGING,
  debugApollo: DEBUG_APOLLO,
}).listen(PORT, () => {
  config({
    default_node_env: 'development',
    purge_dotenv: true,
  })

  console.table({
    __dirname,
    cwd: process.cwd(),
    NODE_ENV: process.env.NODE_ENV,
    LOGGING,
    TRACING,
    DEBUG_APOLLO,
    PORT,
    'Database URL': truthyToMaskedString(process.env.DATABASE_URL),
    'JWT Secret': truthyToMaskedString(process.env.JWT_SECRET),
    'Database URL[]': truthyToMaskedString(process.env['DATABASE_URL']),
    'JWT Secret[]': truthyToMaskedString(process.env['JWT_SECRET']),
  })

  assertEnvVarsExist(['DATABASE_URL', 'JWT_SECRET'])

  console.log(`Server running on http://localhost:${PORT}`)
})
