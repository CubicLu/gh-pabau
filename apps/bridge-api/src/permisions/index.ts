import { allow, rule, shield } from 'graphql-shield'
import { Context } from '../context'

const rules = {
  isAuthenticated: rule('isAuthenticated')(
    async (root, args, ctx: Context): Promise<boolean> => {
      try {
        if (ctx.req.authenticatedUser) return true
      } catch (error) {
        console.log(error)
        return false
      }
    }
  ),
  sameCompany: rule('sameCompany')(
    async (root, args, ctx: Context, info): Promise<boolean> => {
      try {
        if (
          info.returnType.toString().startsWith('[') ||
          info.operation.name.value.includes('aggregate')
        )
          args.where = {
            ...args.where,
            company_id: { equals: ctx.req.authenticatedUser.company },
          }
        return true
      } catch (error) {
        console.error(error)
      }
    }
  ),
  interceptMutation: rule('interceptMutation')(
    async (root, args, ctx: Context, info): Promise<boolean> => {
      try {
        if (
          info.operation.name.value.includes('add') &&
          !args.data?.company?.connect?.id
        ) {
          args.data = {
            ...args.data,
            company: { connect: { id: ctx.req.authenticatedUser.company } },
          }
        }
        return true
      } catch (error) {
        console.error(error)
      }
    }
  ),
}
export const permissions = shield({
  Mutation: {
    login: allow,
    logout: rules.isAuthenticated,
    '*': rules.isAuthenticated && rules.interceptMutation,
  },
  Query: {
    '*': rules.isAuthenticated && rules.sameCompany,
    marketingSourcesCount: rules.isAuthenticated && rules.sameCompany,
    me: rules.isAuthenticated,
    ping: allow,
  },
})
