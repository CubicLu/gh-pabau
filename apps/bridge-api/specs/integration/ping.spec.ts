import { createTestClient } from 'apollo-server-integration-testing'
import { server } from '../../src/app'

const { query } = createTestClient({
  apolloServer: server,
})

it('server responds to ping', () => {
  expect(query(`{ ping }`)).resolves.toMatchSnapshot()
})
