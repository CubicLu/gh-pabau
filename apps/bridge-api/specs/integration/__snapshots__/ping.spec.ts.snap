// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`server responds to ping 1`] = `
Object {
  "data": Object {
    "ping": "pong",
  },
}
`;
